### PR TITLE
[feature-wip](nereids) Support having clause

### PR DESCRIPTION
**Pull request by adonis0147**
# Proposed changes

Implement the having clause for Nereids Planner.

**NOTE:**

This pr aims at making Nereids Planner generate the correct logical plan and physical plan only. The runtime correctness is not the goal in this pr due to `group by` is not ready in Nereids Planner.

## Problem summary

~~Describe your changes.~~

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 1]
```diff
@@ -949,4 +958,18 @@ public Expression visitExist(ExistContext context) {
     public List<Expression> withInList(PredicateContext ctx) {
         return ctx.expression().stream().map(this::getExpression).collect(ImmutableList.toImmutableList());
     }
+
+    static class SelectQuerySpecificationContext {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:962]: if u need it, the better way to do this is add a new Rule include `whereClause? aggClause? havingClause?` named `SelectQuerySpecification` in `DorisParser.g4`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java [hunk 2]
```diff
@@ -155,9 +157,15 @@ private <E extends Expression> List<E> bind(List<E> exprList, List<Plan> inputs,
     }
 
     private <E extends Expression> E bind(E expr, List<Plan> inputs, Plan plan, CascadesContext cascadesContext) {
-        List<Slot> boundedSlots = inputs.stream()
-                .flatMap(input -> input.getOutput().stream())
-                .collect(Collectors.toList());
+        List<NamedExpression> boundedSlots = inputs.stream()
+                .flatMap(input -> {
+                    Plan subPlan = ((GroupPlan) input).getGroup().getLogicalExpression().getPlan();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java:162]: i don't think this is right, because we could write sql like that
```sql
SELECT k1, COUNT(v1) FROM t1 GROUP BY k1 HAVING SUM(v2) > 10;
```
the right way to bind having is write another rule that split is predicates into three parts
1. AggregateFunctions **DO NOT** appear in child aggregate node
2. AggregateFunctions **DO** appear in child aggregate node
3. other expression

And then, 
1. push down the **first part** AggregateFunctions to aggregate node
2. replace **both first part and second part** AggregateFunctions by SlotReferences generated from aggregate node's output expression
3. if the first part is not empty, we need add a project node **on the top** of the filter node transformed from Having

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+import java.util.stream.Stream;
+
+/**
+ * Resolve having clause to the aggregation.
+ */
+public class ResolveHaving extends OneAnalysisRuleFactory {
+    @Override
+    public Rule build() {
+        return RuleType.RESOLVE_HAVING.build(
+            logicalHaving().thenApply(ctx -> {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:54]: ditto, use `logicalHaving(logicalAggregate())` to get what u want

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                        .getLogicalExpression().getPlan();
+                Set<Expression> groupByExpressions = new HashSet<>(aggregate.getGroupByExpressions());
+                Set<Expression> tokens = extractTokensForAggregation(having.getPredicates()).collect(
+                        Collectors.toSet());
+                tokens.forEach(token -> {
+                    if (token instanceof SlotReference && !groupByExpressions.contains(token)
+                            || (token instanceof AggregateFunction
+                                && checkWhetherNestedAggregateFunctionsExist((AggregateFunction) token))) {
+                        throw new AnalysisException("Invalid having clause, " + having.getPredicates());
+                    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:66]: 1. group by key could has Alias
2. Aggregate Function does not exists in LogicalAggregate is OK.
such as below query could run on TPC-H table lineitem
```sql
select l_suppkey + 1, sum(l_linenumber), sum(l_partkey) as b from lineitem group by l_suppkey + 1 having count(l_suppkey) > 5;
```
so, what we do is
1. split all predicates into two parts: has Aggregate Function and do not
2. for has part:
2.1 if we could not find any expression subtree in Aggregate Output, add it to aggregate output
2.2 use aggregate output's SlotReference to replace Expression subtree in predicates
3. for do not have part:
3.1 if we could not find any expression subtree in Aggregate Output or Group Key, throw exception
3.2 if we could find any expression subtree in Group Key but do not in Aggregate Output, add this group by key to aggregate output
3.3 use aggregate output's SlotReference to replace Expression subtree in predicates

**Comment by morrySnow** [comment]
please add some analyze unit test and regression test

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Alias.java [hunk 5]
```diff
@@ -49,7 +49,7 @@ public Alias(Expression child, String name) {
     }
 
     @VisibleForTesting
-    Alias(ExprId exprId, Expression child, String name) {
+    public Alias(ExprId exprId, Expression child, String name) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Alias.java:52]: unnecessary change

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/AbstractPlan.java [hunk 6]
```diff
@@ -69,9 +69,10 @@ public Optional<GroupExpression> getGroupExpression() {
 
     @Override
     public boolean canBind() {
-        return !bound()
+        boolean result = !bound()
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/AbstractPlan.java:72]: unnecessary change

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHaving.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        if (!(object instanceof LogicalHaving) || getClass() != object.getClass()) {
+            return false;
+        }
+        LogicalHaving<CHILD_TYPE> other = (LogicalHaving<CHILD_TYPE>) object;
+        return predicates.equals(other.predicates);
+    }
+
+    @Override
+    public String toString() {
+        return "LogicalHaving (" + predicates + ")";
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHaving.java:109]: we have a new Util function to generate value for toString: https://github.com/apache/doris/blob/a49bde8a718ec96f107ac7d44dda8a9017f7ebc4/fe/fe-core/src/main/java/org/apache/doris/nereids/util/Utils.java#L131

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHaving.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    @Override
+    public boolean equals(Object object) {
+        if (this == object) {
+            return true;
+        }
+        if (!(object instanceof LogicalHaving) || getClass() != object.getClass()) {
+            return false;
+        }
+        LogicalHaving<CHILD_TYPE> other = (LogicalHaving<CHILD_TYPE>) object;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHaving.java:103]: other's child type maybe different with this: so
```suggestion
        LogicalHaving other = (LogicalHaving) object;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+                                return false;
+                            }))) {
+                        throw new AnalysisException(token.toSql() + " in having clause should be grouped by.");
+                    } else if (token instanceof AggregateFunction
+                            && checkWhetherNestedAggregateFunctionsExist((AggregateFunction) token)) {
+                        throw new AnalysisException("Aggregate functions in having clause can't be nested: "
+                                + token.toSql() + ".");
+                    }
+                });
+                return createLogicalFilter(having, tokens);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:79]: if we put some expression to child's output expression, we need add a project on filter to keep the output constant

File: fe/fe-core/src/test/java/org/apache/doris/nereids/parser/HavingClauseTest.java [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+import org.apache.doris.nereids.types.IntegerType;
+import org.apache.doris.nereids.util.FieldChecker;
+import org.apache.doris.nereids.util.PatternMatchSupported;
+import org.apache.doris.nereids.util.PlanChecker;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import org.junit.jupiter.api.Test;
+
+public class HavingClauseTest extends AnalyzeCheckTestBase implements PatternMatchSupported {
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/parser/HavingClauseTest.java:42]: please add a case like this:
```sql
SELECT pk + 1, pk + 1 + 1, pk + 2, SUM(a1), COUNT(a1) + 1, SUM(a1 + a2), COUNT(a2) as v1 GROUP BY pk, pk + 1 HAVING pk > 0 AND COUNT(a1) + 1 > 0 AND SUM(a1 + a2) + 1 > 0 AND v1 + 1 > 0 AND v1 > 0
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java.orig [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * BindSlotReference.
+ */
+public class BindSlotReference implements AnalysisRuleFactory {
```
[the remark this answers is not in the document]
Comment (adonis0147) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java.orig]: Oops. I deleted it.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java [hunk 12]
```diff
@@ -123,11 +126,23 @@ public List<Rule> buildRules() {
...
-
+            RuleType.BINDING_HAVING_SLOT.build(
+                logicalHaving(logicalAggregate()).thenApply(ctx -> {
+                    LogicalHaving<LogicalAggregate<GroupPlan>> having = ctx.root;
+                    LogicalAggregate<GroupPlan> aggregate = having.child();
+                    Set<Slot> boundSlots = Stream.concat(Stream.of(aggregate), aggregate.children().stream())
+                            .flatMap(plan -> plan.getOutput().stream())
+                            .collect(Collectors.toSet());
+                    Expression boundPredicates = new SlotBinder(
+                            toScope(new ArrayList<>(boundSlots)), having, ctx.cascadesContext
```
[the remark this answers is not in the document]
Comment (adonis0147) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java:139]: We should deduplicate the slots, otherwise the binding process will fail due to the ambiguous slots exist.

I added a comment.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Alias.java (hunk 5 above)
Comment (adonis0147) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Alias.java:52]: We should make this function `public` here, due to the package of `org.apache.doris.nereids.parser.HavingCluaseTest` is different from the one of `org.apache.doris.nereids.trees.expressions.Alias`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Resolve having clause to the aggregation.
+ */
+public class ResolveHaving extends OneAnalysisRuleFactory {
```
[the remark this answers is not in the document]
Comment (adonis0147) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:50]: When this rule is applied, all the slots and functions in having clause have been bound. That is the reason I chose the other name.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+                    generateAliasForNewOutputSlots(expression);
+                } else {
+                    for (Expression child : expression.children()) {
+                        resolve(child);
+                    }
+                }
+            }
+        }
+
+        private Pair<Optional<Expression>, Boolean> lookUp(Expression expression) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:108]: add comment to explain the return pair

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+                    .findFirst();
+            if (found.isPresent()) {
+                return Pair.of(found, true);
+            }
+
+            found = groupByExpressions.stream().filter(source -> isEquivalent(source, expression)).findFirst();
+            return Pair.of(found, false);
+        }
+
+        private boolean isEquivalent(Expression source, Expression expression) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:121]: add some comment to explain the meaning of equality

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+
+        private boolean checkWhetherNestedAggregateFunctionsExist(AggregateFunction function) {
+            Predicate<Expression> hasAggregateFunctions = new Predicate<Expression>() {
+                @Override
+                public boolean test(Expression expression) {
+                    return expression instanceof AggregateFunction
+                            || expression.children().stream().anyMatch(this);
+                }
+            };
+            return function.children().stream().anyMatch(hasAggregateFunctions);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:139]: could we just use?
```
function.children().stream().anyMatch(c -> c.anyMatch(AggregateFunction.class::isInstance));
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+        private final List<Expression> groupByExpressions;
+        private final Map<Expression, Slot> substitution = Maps.newHashMap();
+        private final List<NamedExpression> newOutputSlots = Lists.newArrayList();
+
+        Resolver(LogicalAggregate<? extends Plan> aggregate) {
+            outputExpressions = aggregate.getOutputExpressions();
+            groupByExpressions = aggregate.getGroupByExpressions();
+        }
+
+        public void resolve(Expression expression) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:76]: the function has a lot of branch, so we need add comment to explain all the situations it can handle

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+            return newOutputSlots;
+        }
+    }
+
+    private Plan createPlan(LogicalHaving<LogicalAggregate<GroupPlan>> having, Resolver resolver) {
+        LogicalAggregate<GroupPlan> aggregate = having.child();
+        Expression newPredicates = substitute(having.getPredicates(), resolver.getSubstitution());
+        List<NamedExpression> newOutputExpressions = Streams.concat(aggregate.getOutputExpressions().stream(),
+                resolver.getNewOutputSlots().stream()).collect(
+                Collectors.toList());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:162]:
```suggestion
        List<NamedExpression> newOutputExpressions = Streams
                .concat(aggregate.getOutputExpressions().stream(),resolver.getNewOutputSlots().stream())
                .collect(Collectors.toList());
```

Comment (adonis0147) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java]: Refined.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java (hunk 17 above)
Comment (adonis0147) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:76]: Added.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java (hunk 15 above)
Comment (adonis0147) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:143]: Added.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java (hunk 14 above)
Comment (adonis0147) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:124]: Added.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java (hunk 16 above)
Comment (adonis0147) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:161]: We need recursion here otherwise we can't cover the case `sum(a1 + a2 + sum(a2))` which is parsed as `new Sum(new Add(new Add(a1, a2), new Sum(a2)))`

Comment (adonis0147) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java]: Ok, the modification also works. I use this way.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+        return resolver.getNewOutputSlots().isEmpty()
+                ? filter
+                : new LogicalProject<>(
+                        aggregate.getOutputExpressions().stream().map(NamedExpression::toSlot)
+                                .collect(Collectors.toList()),
+                        filter
+                );
+    }
+
+    private Expression substitute(Expression expression, Map<Expression, Slot> substitution) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:188]: org.apache.doris.nereids.trees.expressions.visitor.ExpressionReplacer could do same thing.
```java
ExpressionReplacer.INSTANCE.visit(expression, substitution);
```
u could change `ExpressionReplacer` to
```java
extends DefaultExpressionRewriter<Map<? extends Expression, ? extends Expression>>
```
to avoid change substitution map's generic type.

**Comment by morrySnow** [comment]
please add some regression test under regression-test/suites/nereids_syntax_p0

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java (hunk 19 above)
Comment (adonis0147) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java:189]: Done

**Review by morrySnow — APPROVED** [review]
